aten/src/ATen/native/cpu/Activation.cpp

```diff
@@ -757,12 +762,12 @@ void softshrink_kernel(TensorIteratorBase& iter, const Scalar& lambd) {
     cpu_kernel_vec(
       iter,
       [=](scalar_t a) -> scalar_t {
-        return a > lambd_val ? a - lambd_val : (a < -lambd_val ? a + lambd_val : scalar_t(0));
+        return a > lambd_val ? a - lambd_val : (a < -lambd_val ? a + lambd_val : a * scalar_t(0));
       },
       [=](Vectorized<scalar_t> self_val) -> Vectorized<scalar_t> {
           Vectorized<scalar_t> self_val_t0, self_val_t1;
-          self_val_t0 = (self_val > lambdVec) & (self_val - lambdVec);
-          self_val_t1 = (self_val < -lambd_val) & (self_val + lambdVec);
+          self_val_t0 = ((self_val > lambdVec) | (self_val.isnan())) & (self_val - lambdVec);
+          self_val_t1 = ((self_val < -lambd_val) | (self_val.isnan())) & (self_val + lambdVec);
```

Comment:
The code will propagate NaN values correctly. Previously, the comparison self_val > lambdVec always returned False when the input was NaN, because any comparison with NaN evaluated to False. This meant self_val - lambdVec wasn't propagating NaN values and instead defaulted to 0. The mask now will properly detect NaN inputs allowing self_val - lambdVec to return NaN (since + or - op with NaN results in NaN).

Reply:
Your explanation makes sense, do you know where the changes to the vectorized path are tested?

Reply:
There are tests which compares the output of the compiled softshrink with the functional version(with just basic ops). Functional version is here, which I also changed with multiplication to get nan value propagated:

https://github.com/pytorch/pytorch/blob/643b337c02fd836f7a118e63d25cccd05ccc3bdf/torch/_refs/nn/functional/__init__.py#L501-L516

From what I understood from reading the code, the functional version is run with different input shapes as well as different input devices and dtypes(selected by `PYTORCH_OPINFO_SAMPLE_INPUT_INDEX`). Shapes are defined in core.py of opinfo:

https://github.com/pytorch/pytorch/blob/723498aab8803702dfd508be05a1ebe525112ebd/torch/testing/_internal/opinfo/core.py#L1987-L1999

One such test is:

`PYTORCH_OPINFO_SAMPLE_INPUT_INDEX=14 PYTORCH_TEST_WITH_ASAN=1 PYTORCH_TEST_WITH_UBSAN=1 python test/test_ops.py TestCommonCPU.test_python_ref_torch_fallback__refs_nn_functional_softshrink_cpu_float32`

or for cuda and float16(just another example):

`PYTORCH_OPINFO_SAMPLE_INPUT_INDEX=14 python test/test_ops.py TestCommonCUDA.test_python_ref_torch_fallback__refs_nn_functional_softshrink_cuda_float16`

Reply:
Amazing, thank you for clarifying, and to double check, these sample inputs have nans in them?